### PR TITLE
[SPARK-30382][SQL] Remove Hive LogUtils usage to prevent ClassNotFoundException

### PR DESCRIPTION
**Pull request by ajithme**
### What changes were proposed in this pull request?
Avoid hive log initialisation as https://github.com/apache/hive/blob/rel/release-2.3.5/common/src/java/org/apache/hadoop/hive/common/LogUtils.java introduces dependency over `org.apache.logging.log4j.core.impl.Log4jContextFactory` which is missing in our spark installer classpath directly. I believe the `LogUtils.initHiveLog4j()` code is here as the HiveServer2 class is copied from Hive.

### Why are the changes needed?
To make `start-thriftserver.sh --help` command success.

Currently, start-thriftserver.sh --help throws
```
...
Thrift server options:
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/logging/log4j/spi/LoggerContextFactory
	at org.apache.hive.service.server.HiveServer2.main(HiveServer2.java:167)
	at org.apache.spark.sql.hive.thriftserver.HiveThriftServer2$.main(HiveThriftServer2.scala:82)
	at org.apache.spark.sql.hive.thriftserver.HiveThriftServer2.main(HiveThriftServer2.scala)
Caused by: java.lang.ClassNotFoundException: org.apache.logging.log4j.spi.LoggerContextFactory
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 3 more
```

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Checked Manually

### CODE REVIEW

**Comment by ajithme** [comment]
cc @srowen @wangyum @dongjoon-hyun @cloud-fan

**Comment by srowen** [comment]
So this only affects the 2.x code that is copied?
Yes, it's a copy, so we ideally don't want to modify it. I'm not sure what the implications are of not initializing the logging, right?
Is it not a matter of making sure the script or build includes log4j?

**Comment by ajithme** [comment]
> So this only affects the 2.x code that is copied?
> Yes, it's a copy, so we ideally don't want to modify it. I'm not sure what the implications are of not initializing the logging, right?
> Is it not a matter of making sure the script or build includes log4j?

I see HiveServer2.main being called only to print help message.

**Comment by ajithme** [comment]
> So this only affects the 2.x code that is copied?
> Yes, it's a copy, so we ideally don't want to modify it. I'm not sure what the implications are of not initializing the logging, right?
> Is it not a matter of making sure the script or build includes log4j?

Currently HiveServer2.main is called only to print help message, all other operations currently work without Hive log initialization. Please correct if wrong

**Comment by srowen** [comment]
That I don't know. @wangyum @steveloughran maybe, do you have any insight into what this is used for?

**Comment by HeartSaVioR** [comment]
Looks like the problem is that Hive doesn't only leverage slf4j, but also leverages Log4j"2" API to programmatically configuring Log4j via configured in Hive conf. If I understand correctly, Spark uses Log4j 1, and that is also applied to hive-thriftserver module (since it should run with Spark dependencies). 

We also seem to exclude everything for log4j2 as it conflicts with log4j 1:

```
          <!-- org.apache.logging.log4j:* conflict with log4j 1.2.17 -->
          <exclusion>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>*</artifactId>
          </exclusion>
```

**Comment by ajithme** [comment]
Thanks for the details @HeartSaVioR So what is the suggestion here, should we 
1. include `log4j2` to complete the dependency OR
2. drop the code which leverages on log4j2, 

personally after tracing the code i see log4j2 for hive is only initalized in help flow so it should be safe to delete the `LogUtils.initHiveLog4j()` , hence the PR follows approach2

File: sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java [hunk 1]
```diff
@@ -158,25 +156,18 @@ private static void startHiveServer2() throws Throwable {
...
-      ServerOptionsProcessorResponse oprocResponse = oproc.parse(args);
+    ServerOptionsProcessor oproc = new ServerOptionsProcessor("hiveserver2");
+    ServerOptionsProcessorResponse oprocResponse = oproc.parse(args);
 
-      // NOTE: It is critical to do this here so that log4j is reinitialized
-      // before any of the other core hive classes are loaded
-      String initLog4jMessage = LogUtils.initHiveLog4j();
-      LOG.debug(initLog4jMessage);
-      HiveStringUtils.startupShutdownMessage(HiveServer2.class, args, LOG);
+    // NOTE: It is critical to do this here so that log4j is reinitialized
```
Comment (srowen) [inline sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java]: Delete this comment?

**Review by srowen — COMMENTED** [review]
I doubt we can include log4j 2 successfully. Hm. So this code path doesn't get used unless --help is called? can you summarize why, because this looks like a main entry point. I don't know this code at all though. I'm worried it will break other logging.

**Comment by ajithme** [comment]
> I doubt we can include log4j 2 successfully. Hm. So this code path doesn't get used unless --help is called? can you summarize why, because this looks like a main entry point. I don't know this code at all though. I'm worried it will break other logging.

Sure, currently for spark thrift server, the entrypoint class is `org.apache.spark.sql.hive.thriftserver.HiveThriftServer2` https://github.com/apache/spark/blob/master/sbin/start-thriftserver.sh#L32

In `org.apache.spark.sql.hive.thriftserver.HiveThriftServer2#main`, we can see, its calling hive's class `org.apache.hive.service.server.HiveServer2#main` only incase of argument passed to the HiveThriftServer2#main is `--help` or `-h`, referring to code at https://github.com/apache/spark/blob/master/sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala#L81 which inturn invokes our culprit `LogUtils.initHiveLog4j()`

So, Any other flow which is not passing help argument to `org.apache.spark.sql.hive.thriftserver.HiveThriftServer2`, it will not invoke `org.apache.hive.service.server.HiveServer2#main`

File: sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java (hunk 1 above)
Comment (ajithme) [reply sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java]: Done

**Comment by srowen** [comment]
OK I think I buy it.

**Comment by srowen** [comment]
Jenkins retest this please

**Review by HeartSaVioR — APPROVED** [review]
LGTM assuming manually tested, and we are OK to modify included hive code.

File: sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java [hunk 2]
```diff
@@ -158,25 +156,16 @@ private static void startHiveServer2() throws Throwable {
...
-      // before any of the other core hive classes are loaded
-      String initLog4jMessage = LogUtils.initHiveLog4j();
-      LOG.debug(initLog4jMessage);
-      HiveStringUtils.startupShutdownMessage(HiveServer2.class, args, LOG);
+    HiveStringUtils.startupShutdownMessage(HiveServer2.class, args, LOG);
 
-      // Log debug message from "oproc" after log4j initialize properly
-      LOG.debug(oproc.getDebugMessage().toString());
+    // Log debug message from "oproc" after log4j initialize properly
+    LOG.debug(oproc.getDebugMessage().toString());
```
Comment (cloud-fan) [inline sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java]: is it OK to call `LOG.debug` without initialization?

Comment (ajithme) [reply sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java]: It will ignore the log message, but not throw any exception. As we see here, we initialise spark loggers after this call, https://github.com/apache/spark/blob/master/sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala#L87 
So calling logger before this (even using commons.logging) will not print anything in logs as it doesn't have any appenders

Comment (steveloughran) [reply sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java]: the toString() could trigger exceptions (NPE etc) If you aren't careful and there is the extra cost of that operation. Its why scala string expansion and SLF4J {} entries are better

Comment (cloud-fan) [reply sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java]: shall we just remove this line if it's noop?

Comment (ajithme) [reply sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java]: @steveloughran as i see the code `toString()` is over a `StringBuilder` which is always initialised in `org.apache.hive.service.server.HiveServer2.ServerOptionsProcessor#debugMessage` and as also this is java file dint use scala string expansion. Please correct me if wrong.

@cloud-fan removed as per suggestion

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @ajithme . Thank you for your contribution. I have two comments.

1. This PR fixes only Hive 2.3 code path. We need to remove Hive 1.2 code path. We want to keep both source code path identical as much as possible.
```
$ git grep 'org.apache.hadoop.hive.common.LogUtils'
sql/hive-thriftserver/v1.2/src/main/java/org/apache/hive/service/server/HiveServer2.java:import org.apache.hadoop.hive.common.LogUtils;
sql/hive-thriftserver/v1.2/src/main/java/org/apache/hive/service/server/HiveServer2.java:import org.apache.hadoop.hive.common.LogUtils.LogInitializationException;
sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java:import org.apache.hadoop.hive.common.LogUtils;
sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java:import org.apache.hadoop.hive.common.LogUtils.LogInitializationException;
```

2. It would be better describe your PR's content in the PR title. You don't need to copy the JIRA title. I updated a little for you at this time. Also, I updated the PR description, too.

**Comment by ajithme** [comment]
> Hi, @ajithme . Thank you for your contribution. I have two comments.
> 
> 1. This PR fixes only Hive 2.3 code path. We need to remove Hive 1.2 code path. We want to keep both source code path identical as much as possible.
> 
> ```
> $ git grep 'org.apache.hadoop.hive.common.LogUtils'
> sql/hive-thriftserver/v1.2/src/main/java/org/apache/hive/service/server/HiveServer2.java:import org.apache.hadoop.hive.common.LogUtils;
> sql/hive-thriftserver/v1.2/src/main/java/org/apache/hive/service/server/HiveServer2.java:import org.apache.hadoop.hive.common.LogUtils.LogInitializationException;
> sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java:import org.apache.hadoop.hive.common.LogUtils;
> sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java:import org.apache.hadoop.hive.common.LogUtils.LogInitializationException;
> ```
> 
> 1. It would be better describe your PR's content in the PR title. You don't need to copy the JIRA title. I updated a little for you at this time. Also, I updated the PR description, too.

@dongjoon-hyun Thanks for the inputs. I have updated the PR to Handle for Hive 1.2. Also thanks for updating PR details, will keep in mind next time.

**Comment by srowen** [comment]
Jenkins retest this please

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master. 
Thank you, @ajithme and all.

File: sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java [hunk 3]
```diff
@@ -158,25 +156,13 @@ private static void startHiveServer2() throws Throwable {
...
-      // Log debug message from "oproc" after log4j initialize properly
-      LOG.debug(oproc.getDebugMessage().toString());
-
-      // Call the executor which will execute the appropriate command based on the parsed options
-      oprocResponse.getServerOptionsExecutor().execute();
-    } catch (LogInitializationException e) {
-      LOG.error("Error initializing log: " + e.getMessage(), e);
-      System.exit(-1);
-    }
+    // Call the executor which will execute s  appropriate command based on the parsed options
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/v2.3/src/main/java/org/apache/hive/service/server/HiveServer2.java]: I will fix this line during merging. This line seems to be changed accidentaly.
